### PR TITLE
🧹 Better handling of exceptions in `get_email_code` 

### LINKED ISSUE #362: bug: :bug: get_email_code hides exceptions when they happen
The convention used throughout the function `get_email_code` to just return an empty string if there is an error hides excpetions. I would suggest that we explicitly raise exceptions when errors happen else the sign workflow will just ignore the problem and fail later when signing in. I think that makes debugging issues harder.

_Originally posted by @jaron-l in https://github.com/mintapi/mintapi/pull/360#discussion_r763277466_

### PR DESCRIPTION
**Pull request by burkematthew**
* Refactored the `sign_in` method for a more modular approach
* Changing `logger.error` instances to exceptions in `get_email_code`

Resolves #362

### CODE REVIEW

File: mintapi/signIn.py
Comment (mrooney) [inline mintapi/signIn.py:65]: Sometimes we add `rv` to the error message and sometimes not. Would it make sense to always include it like on line 61? I realize you are just converting them as-is; I'm just curious if it can be improved even a little more :)

File: mintapi/signIn.py
Comment (mrooney) [inline mintapi/signIn.py:74]: Is `num` actually meaningful to the user? Would `rv` be better here? I'm not actually sure what it looks like when it isn't "OK" or if `data` is better.

**Comment by mrooney** [comment]
Nice! Does this also handle the issue of making the login flow more dynamic based on what is being prompted for?

**Comment by burkematthew** [comment]
@mrooney no, not yet, though it does make that project clearer. I wanted to get this done first before I made the flow more dynamic.

File: mintapi/signIn.py
Comment (burkematthew) [reply mintapi/signIn.py:74]: That's a good question. `rv` is most likely the most meaningful and we can standardize the inclusion of that in each message.

File: mintapi/signIn.py
Comment (burkematthew) [reply mintapi/signIn.py:65]: Yup, I can do that.

**Comment by burkematthew** [comment]
@mrooney back over to you!  Once this merges, I'll work on the more dynamic sign-in flow.

**Comment by mrooney** [comment]
@burkematthew okay, I was having trouble following the diff, but I think I am understanding it better on a second read. So those changes aren't related to the PR title directly, but is basically breaking up one super long function into little parts, that are then called in the same order as before? And in the future, we can add more detection to decide when to call each one?

Are there any potential regressions you could imagine with this? It seems hard to test manually because we can't necessarily control which flow/UI we get 100%, but were you able to test it with the various switches we can provide?

**Comment by burkematthew** [comment]
@mrooney yes I know it made the main change a little more unclear, but I thought it helped understand where exceptions could arise in the overall login flow. 

I will go through some more testing iterations to verify, but it all worked without issue when I was working on this. And really there shouldn't be any regressions unless I just copied/compartmentalized the flow wrong. Again, I'll go through a few more times to verify it's all as expected.

**Review by mrooney — DISMISSED** [review]
Got it, thanks for the clarification! In that case, +1 from me once you are done any final checks you want to do!

**Comment by burkematthew** [comment]
@mrooney all the tests I ran looked good; no issues.

@jaron-l anything else you want to add here?

**Comment by burkematthew** [comment]
FYI, once we get this merged, I'll go ahead and start on the more dynamic MFA handling.

**Review by mrooney — APPROVED** [review]

**Comment by mrooney** [comment]
@burkematthew got it, LGTM, approved and merged! I want to unblock you but if @jaron-l has any concerns please do bring them up and we can address them in the next PR :)

**Pull request merged**
